src/com/facebook/buck/maven/Resolver.java

```diff
@@ -98,75 +119,116 @@
-    this.repos = builder.build();
-
-    this.locator = AetherUtil.initServiceLocator();
+    for (String artifact : config.artifacts) {
+      if (artifact.endsWith(".pom")) {
+        Model model = loadPomModel(Paths.get(artifact));
+        repoBuilder.addAll(getReposFromPom(model));
+        for (Dependency dep : getDependenciesFromPom(model)) {
+          dependencyBuilder.put(buildKey(dep.getArtifact()), dep);
+        }
+      } else {
+        Dependency dep = getDependencyFromString(artifact);
+        dependencyBuilder.put(buildKey(dep.getArtifact()), dep);
+      }
+    }
+    this.repos = repoBuilder.build();
+    this.specifiedDependencies = dependencyBuilder.build();
   }
 
-  public void resolve(String... mavenCoords) throws RepositoryException, IOException {
-    RepositorySystem repoSys = locator.getService(RepositorySystem.class);
-    RepositorySystemSession session = newSession(repoSys);
-
-    Set<String> specifiedArtifacts = new HashSet<>();
-    for (String coords : mavenCoords) {
-      specifiedArtifacts.add(coords);
-    }
+  public void resolve() throws RepositoryException, ExecutionException, InterruptedException,
+      IOException {
+    final RepositorySystem repoSys = locator.getService(RepositorySystem.class);
+    final RepositorySystemSession session = newSession(repoSys);
 
     ImmutableMap<String, Artifact> knownDeps = getRunTimeTransitiveDeps(
         repoSys,
         session,
-        mavenCoords);
+        specifiedDependencies.values());
 
     // We now have the complete set of dependencies. Build the graph of dependencies. We'd like
     // aether to do this for us, but it doesn't preserve the complete dependency information we need
     // to accurately construct build files.
-    MutableDirectedGraph<Artifact> graph = buildDependencyGraph(repoSys, session, knownDeps);
+    final MutableDirectedGraph<Artifact> graph = buildDependencyGraph(repoSys, session, knownDeps);
 
     // Now we have the graph, grab the sources and jars for each dependency, as well as the relevant
     // checksums (which are download by default. Yay!)
+    ImmutableSetMultimap<Path, Prebuilt> downloadedArtifacts =
+        downloadArtifacts(repoSys, session, graph);
 
-    Map<Path, SortedSet<Prebuilt>> buckFiles = new HashMap<>();
+    createBuckFiles(downloadedArtifacts);
+  }
 
-    for (Artifact artifact : graph.getNodes()) {
-      List<String> extraVisibility = specifiedArtifacts.contains(artifact.toString()) ?
-        this.visibility : ImmutableList.<String>of();
-      downloadArtifact(artifact, repoSys, session, buckFiles, graph, extraVisibility);
-    }
+  private ImmutableSetMultimap<Path, Prebuilt> downloadArtifacts(
+      final RepositorySystem repoSys,
+      final RepositorySystemSession session,
+      final MutableDirectedGraph<Artifact> graph) throws ExecutionException, InterruptedException {
+    ListeningExecutorService exec = MoreExecutors.listeningDecorator(
+        Executors.newWorkStealingPool());
```

Comment:
to make it easier to debug we tend to use the MoreExecutors wrappers which give the threads in the pool useful names. Any chance you could update this to name the threads something like "MavenArtifactDownload"?